### PR TITLE
scheduler_perf: code cleanups

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This is in preparation of https://github.com/kubernetes/kubernetes/pull/117196

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @alculquicondor @kerthcet

/hold

For review by @alculquicondor

### CODE REVIEW

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 1]
```diff
@@ -1020,6 +1018,18 @@ func runWorkload(ctx context.Context, b *testing.B, tc *testCase, w *workload) [
 	return dataItems
 }
 
+func maybeCreateNamespace(ctx context.Context, b *testing.B, client clientset.Interface, namespace string, podsPerNamespace *map[string]int) {
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1021]: For clarity: `@thockin` asked me to make that same change in some other code to make it more obvious that the map is being modified by a function.

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 2]
```diff
@@ -750,14 +750,20 @@ func runWorkload(ctx context.Context, b *testing.B, tc *testCase, w *workload) [
 			b.Fatalf("validate scheduler config file failed: %v", err)
 		}
 	}
-	podInformer, client, dynClient := mustSetupScheduler(ctx, b, cfg)
+	informerFactory, client, dynClient := mustSetupScheduler(ctx, b, cfg)
+
+	// Additional informers needed for testing.
+	podInformer := informerFactory.Core().V1().Pods()
+	informerFactory.Start(ctx.Done())
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:757]: `Start` can be called more than once and has to be called again after creating additional informers.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1021]: Is this a best practice for golang? Or we hope to follow this in K8s.

**Review by kerthcet — COMMENTED** [review]
For confirmation: `This is in preparation of https://github.com/kubernetes/kubernetes/pull/117196`, but I didn't see any codes related in that PR.

**Comment by pohly** [comment]
I meant https://github.com/kubernetes/kubernetes/pull/116207 - that needs maybeCreateNamespace for other entities.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1021]: I don't know about Golang in general, but at least `@thockin` wanted it for API validation code.

If you and @alculquicondor disagree, then I can remove the pointer here.

[the remark this answers is not in the document]
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1021]: Just want to dig something I may learn from here.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:757]: I think no need to start again, we didn't add any new informers.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 1 above)
Comment (kerthcet) [inline test/integration/scheduler_perf/scheduler_perf_test.go:1021]: enough?

```suggestion
func createNamespace(ctx context.Context, b *testing.B, client clientset.Interface, namespace string, podsPerNamespace *map[string]int) {
```

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:757]: How do you know that `informerFactory.Core().V1().Pods()` is not a new informer?

If it is not new, then `Start` will not start it again. But if it is new, then not calling `Start` is a mistake. Calling `Start` is therefore the safe, more easily reviewable choice.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 1 above)
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1021]: It doesn't always create the namespace.

`ensureNamespace`?

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:757]: > How do you know that informerFactory.Core().V1().Pods() is not a new informer?

I mean we maintain an in-place pod informer here. And it's a pod informer customized 
which only watches for unterminated pods. If it's a new pod informer here, it doesn't make a lot of sense. https://github.com/kubernetes/kubernetes/blob/3554bcde8730991ec43e620d006899839ddd251d/pkg/scheduler/scheduler.go#L371-L375

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 1 above)
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1021]: I like `createNamespaceIfNotPresent`, but up to you.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:757]: That code is interesting. What is the effect of that `informerFactory.InformerFor` on the generic `informerFactory` instance? `informerFactory.InformerFor` returns a new `SharedIndexInformer`, but that value is not used anywhere.

Regarding the correctness of my code: it's doing the same thing as before:
- `informerFactory := scheduler.NewInformerFactory(clientSet, 0)`
- `podInformer := informerFactory.Core().V1().Pods()`
- `informerFactory.Start(ctx.Done())`
- `informerFactory.WaitForCacheSync(ctx.Done())`

There are some extra calls to `Start` and `WaitForCacheSync` (in `StartScheduler`) but that isn't wrong.

[the remark this answers is not in the document]
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:757]: > What is the effect of that informerFactory.InformerFor on the generic informerFactory instance?

InformerFor will register a new pod informer here if not exists, if exists, return it. The returned value is the exactly pod informer.
So when we call `informerFactory.Core().V1().Pods().Informer()` we'll get the registered pod informer again.

I'm not saying you're wrong, but seems not necessary, we'll get the registered informer from `informerFactory.Core().V1().Pods().Informer()` or `informerFactory.Core().V1().Pods().Lister()`, it's not a new one, it's the scheduler customized pod informer.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 1 above)
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Renamed to `createNamespaceIfNotPresent`.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Got it. Okay, let's drop the Start+WaitForCacheSync. I added a comment explaining why it's not needed.

**Comment by kerthcet** [comment]
/lgtm
/hold

Leave unhold to @alculquicondor

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold cancel

**Pull request merged**